### ISSUE TITLE
No constructor for type ObjectiveTrait, JObject(List())

### ISSUE THREAD

**Issue by wanghao19970205** [dmlc/xgboost#5936]
When I load the PipelineModel, I got error; anyone could help me ?
2020-07-24 10:27:11 [Driver] ERROR [ApplicationMaster:91]: User class threw exception: org.json4s.package$MappingException: No constructor for type ObjectiveTrait, JObject(List())
org.json4s.package$MappingException: No constructor for type ObjectiveTrait, JObject(List())
	at org.json4s.Extraction$ClassInstanceBuilder.org$json4s$Extraction$ClassInstanceBuilder$$constructor(Extraction.scala:417)
	at org.json4s.Extraction$ClassInstanceBuilder.org$json4s$Extraction$ClassInstanceBuilder$$instantiate(Extraction.scala:468)
	at org.json4s.Extraction$ClassInstanceBuilder$$anonfun$result$6.apply(Extraction.scala:515)
	at org.json4s.Extraction$ClassInstanceBuilder$$anonfun$result$6.apply(Extraction.scala:512)
	at org.json4s.Extraction$.org$json4s$Extraction$$customOrElse(Extraction.scala:524)
	at org.json4s.Extraction$ClassInstanceBuilder.result(Extraction.scala:512)
	at org.json4s.Extraction$.extract(Extraction.scala:351)
	at org.json4s.Extraction$.extract(Extraction.scala:42)
	at org.json4s.ExtractableJsonAstNode.extract(ExtractableJsonAstNode.scala:21)
	at ml.dmlc.xgboost4j.scala.spark.params.CustomObjParam.jsonDecode(CustomParams.scala:63)
	at ml.dmlc.xgboost4j.scala.spark.params.CustomObjParam.jsonDecode(CustomParams.scala:47)
	at ml.dmlc.xgboost4j.scala.spark.params.DefaultXGBoostParamsReader$$anonfun$getAndSetParams$1.apply(DefaultXGBoostParamsReader.scala:117)
	at ml.dmlc.xgboost4j.scala.spark.params.DefaultXGBoostParamsReader$$anonfun$getAndSetParams$1.apply(DefaultXGBoostParamsReader.scala:115)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at ml.dmlc.xgboost4j.scala.spark.params.DefaultXGBoostParamsReader$.getAndSetParams(DefaultXGBoostParamsReader.scala:115)
	at ml.dmlc.xgboost4j.scala.spark.XGBoostRegressionModel$XGBoostRegressionModelReader.load(XGBoostRegressor.scala:447)
	at ml.dmlc.xgboost4j.scala.spark.XGBoostRegressionModel$XGBoostRegressionModelReader.load(XGBoostRegressor.scala:431)
	at org.apache.spark.ml.util.DefaultParamsReader$.loadParamsInstance(ReadWrite.scala:453)
	at org.apache.spark.ml.Pipeline$SharedReadWrite$$anonfun$4.apply(Pipeline.scala:274)
	at org.apache.spark.ml.Pipeline$SharedReadWrite$$anonfun$4.apply(Pipeline.scala:272)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
	at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
	at org.apache.spark.ml.Pipeline$SharedReadWrite$.load(Pipeline.scala:272)
	at org.apache.spark.ml.PipelineModel$PipelineModelReader.load(Pipeline.scala:348)
	at org.apache.spark.ml.PipelineModel$PipelineModelReader.load(Pipeline.scala:342)
	at org.apache.spark.ml.util.MLReadable$class.load(ReadWrite.scala:223)
	at org.apache.spark.ml.PipelineModel$.load(Pipeline.scala:332)
	at com.baidu.map.be.streamingphpui.task.scene.TaskScene5Predict$.doPredict(TaskScene5Predict.scala:60)
	at com.baidu.map.be.streamingphpui.task.scene.TaskScene5ArriveSample$.run(TaskScene5ArriveSample.scala:112)
	at com.baidu.map.be.streamingphpui.task.SparkMain$.run(SparkMain.scala:234)
	at com.baidu.map.be.streamingphpui.task.SparkMain$.main(SparkMain.scala:138)
	at com.baidu.map.be.streamingphpui.task.SparkMain.main(SparkMain.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$4.run(ApplicationMaster.scala:774)

**Comment by trivialfis** [comment]
Doesn't seem to be related to XGBoost.

**Comment by wanghao19970205** [comment]
I customized the loss function with reference to the code here https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-example/src/main/scala/ml/dmlc/xgboost4j/scala/example/CustomObjective.scala，It can be trained normally, but this error is reported when the load model to predicts

**Comment by wanghao19970205** [comment]
The following is the code of my custom loss function：
val xgb = new XGBoostRegressor(xgbParams).setCustomObj(Loss)

class Loss extends ObjectiveTrait {
    private val logger: Log = LogFactory.getLog(classOf[FocalLoss])
    /**
     * Focal loss
     * @param predicts untransformed margin predicts
     * @param dtrain   training data
     * @return List with two float array, correspond to first order grad and second order grad
     */
    override def getGradient(predicts: Array[Array[Float]], dtrain: DMatrix): List[Array[Float]] = {
      /**
       * simple sigmoid func
       *
       * @param input
       * @return Note: this func is not concern about numerical stability
       */
      def sigmoid(input: Float): Float = {
        (1 / (1 + Math.exp(-input))).toFloat
      }
      def transform(predicts: Array[Array[Float]]): Array[Array[Float]] = {
        val nrow = predicts.length
        val transPredicts = Array.fill[Float](nrow, 1)(0)
        for (i <- 0 until nrow) {
          transPredicts(i)(0) = sigmoid(predicts(i)(0))
        }
        transPredicts
      }

      val transPredicts = transform(predicts)
      val alpha = 0.75f    //调大提高召回，减少精确率；调小反之
      val gamma_indct = 1.0f   //调大调小都会提升精度，减小召回
      val nrow = predicts.length
      var label: Array[Float] = null
      try {
        label = dtrain.getLabel
      } catch {
        case e: XGBoostError =>
          logger.error(e)
          null
        case _: Throwable =>
          null
      }
      val gradArr = new Array[Float](nrow)
      val hessArr = new Array[Float](nrow)
      val gradients = new ListBuffer[Array[Float]]
      for (i <- 0 until nrow){
        val sigmoid_pred = transPredicts(i)(0)
        val g1 = sigmoid_pred * (1 - sigmoid_pred)
        val g2 = label(i) + math.pow(-1, label(i)) * sigmoid_pred
        val g3 = sigmoid_pred + label(i) - 1
        val g4 = 1 - g2
        val grad = (alpha*label(i) + (1 - alpha)*(1 - label(i)))*(gamma_indct * g3 * math.pow(g2, gamma_indct) * math.log(g4 + 1e-9) + math.pow(-1, label(i)) * math.pow(g2, gamma_indct + 1))
        val hess_1 = math.pow(g2, gamma_indct) + gamma_indct * Math.pow(-1, label(i)) * g3 * math.pow(g2, gamma_indct - 1)
        val hess_2 = math.pow(-1, label(i)) * g3 * math.pow(g2, gamma_indct) / g4
        val hess = (alpha*label(i) + (1 - alpha)*(1 - label(i)))*(((hess_1 * math.log(g4 + 1e-9) - hess_2) * gamma_indct + (gamma_indct + 1) * math.pow(g2, gamma_indct)) * g1)
        gradArr(i) = grad.toFloat
        hessArr(i) = hess.toFloat
      }
      gradients += gradArr
      gradients += hessArr
      gradients.toList
    }
  }

**Comment by hcho3** [comment]
I think it might have to do with a mismatched Scala version? Make sure to use the same Scala version as XGBoost uses.

**Comment by zhitao-wang** [comment]
> When I load the PipelineModel, I got error; anyone could help me ?
> 2020-07-24 10:27:11 [Driver] ERROR [ApplicationMaster:91]: User class threw exception: org.json4s.package$MappingException: No constructor for type ObjectiveTrait, JObject(List())
> org.json4s.package$MappingException: No constructor for type ObjectiveTrait, JObject(List())
> at org.json4s.Extraction$ClassInstanceBuilder.org$json4s$Extraction$ClassInstanceBuilder$$constructor(Extraction.scala:417)
> at org.json4s.Extraction$ClassInstanceBuilder.org$json4s$Extraction$ClassInstanceBuilder$$instantiate(Extraction.scala:468)
> at org.json4s.Extraction$ClassInstanceBuilder$$anonfun$result$6.apply(Extraction.scala:515)
> at org.json4s.Extraction$ClassInstanceBuilder$$anonfun$result$6.apply(Extraction.scala:512)
> at org.json4s.Extraction$.org$json4s$Extraction$$customOrElse(Extraction.scala:524)
> at org.json4s.Extraction$ClassInstanceBuilder.result(Extraction.scala:512)
> at org.json4s.Extraction$.extract(Extraction.scala:351)
> at org.json4s.Extraction$.extract(Extraction.scala:42)
> at org.json4s.ExtractableJsonAstNode.extract(ExtractableJsonAstNode.scala:21)
> at ml.dmlc.xgboost4j.scala.spark.params.CustomObjParam.jsonDecode(CustomParams.scala:63)
> at ml.dmlc.xgboost4j.scala.spark.params.CustomObjParam.jsonDecode(CustomParams.scala:47)
> at ml.dmlc.xgboost4j.scala.spark.params.DefaultXGBoostParamsReader$$anonfun$getAndSetParams$1.apply(DefaultXGBoostParamsReader.scala:117)
> at ml.dmlc.xgboost4j.scala.spark.params.DefaultXGBoostParamsReader$$anonfun$getAndSetParams$1.apply(DefaultXGBoostParamsReader.scala:115)
> at scala.collection.immutable.List.foreach(List.scala:381)
> at ml.dmlc.xgboost4j.scala.spark.params.DefaultXGBoostParamsReader$.getAndSetParams(DefaultXGBoostParamsReader.scala:115)
> at ml.dmlc.xgboost4j.scala.spark.XGBoostRegressionModel$XGBoostRegressionModelReader.load(XGBoostRegressor.scala:447)
> at ml.dmlc.xgboost4j.scala.spark.XGBoostRegressionModel$XGBoostRegressionModelReader.load(XGBoostRegressor.scala:431)
> at org.apache.spark.ml.util.DefaultParamsReader$.loadParamsInstance(ReadWrite.scala:453)
> at org.apache.spark.ml.Pipeline$SharedReadWrite$$anonfun$4.apply(Pipeline.scala:274)
> at org.apache.spark.ml.Pipeline$SharedReadWrite$$anonfun$4.apply(Pipeline.scala:272)
> at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
> at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)
> at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
> at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
> at scala.collection.TraversableLike$class.map(TraversableLike.scala:234)
> at scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)
> at org.apache.spark.ml.Pipeline$SharedReadWrite$.load(Pipeline.scala:272)
> at org.apache.spark.ml.PipelineModel$PipelineModelReader.load(Pipeline.scala:348)
> at org.apache.spark.ml.PipelineModel$PipelineModelReader.load(Pipeline.scala:342)
> at org.apache.spark.ml.util.MLReadable$class.load(ReadWrite.scala:223)
> at org.apache.spark.ml.PipelineModel$.load(Pipeline.scala:332)
> at com.baidu.map.be.streamingphpui.task.scene.TaskScene5Predict$.doPredict(TaskScene5Predict.scala:60)
> at com.baidu.map.be.streamingphpui.task.scene.TaskScene5ArriveSample$.run(TaskScene5ArriveSample.scala:112)
> at com.baidu.map.be.streamingphpui.task.SparkMain$.run(SparkMain.scala:234)
> at com.baidu.map.be.streamingphpui.task.SparkMain$.main(SparkMain.scala:138)
> at com.baidu.map.be.streamingphpui.task.SparkMain.main(SparkMain.scala)
> at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> at java.lang.reflect.Method.invoke(Method.java:498)
> at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$4.run(ApplicationMaster.scala:774)



> The following is the code of my custom loss function：
> 
> ```
> val xgb = new XGBoostRegressor(xgbParams).setCustomObj(Loss)
> 
> class Loss extends ObjectiveTrait {
>     private val logger: Log = LogFactory.getLog(classOf[Loss])
>     /**
>      *loss
>      * @param predicts untransformed margin predicts
>      * @param dtrain   training data
>      * @return List with two float array, correspond to first order grad and second order grad
>      */
>     override def getGradient(predicts: Array[Array[Float]], dtrain: DMatrix): List[Array[Float]] = {
>       /**
>        * simple sigmoid func
>        *
>        * @param input
>        * @return Note: this func is not concern about numerical stability
>        */
>       def sigmoid(input: Float): Float = {
>         (1 / (1 + Math.exp(-input))).toFloat
>       }
>       def transform(predicts: Array[Array[Float]]): Array[Array[Float]] = {
>         val nrow = predicts.length
>         val transPredicts = Array.fill[Float](nrow, 1)(0)
>         for (i <- 0 until nrow) {
>           transPredicts(i)(0) = sigmoid(predicts(i)(0))
>         }
>         transPredicts
>       }
> 
>       val transPredicts = transform(predicts)
>       val alpha = 0.75f    
>       val gamma_indct = 1.0f   
>       val nrow = predicts.length
>       var label: Array[Float] = null
>       try {
>         label = dtrain.getLabel
>       } catch {
>         case e: XGBoostError =>
>           logger.error(e)
>           null
>         case _: Throwable =>
>           null
>       }
>       val gradArr = new Array[Float](nrow)
>       val hessArr = new Array[Float](nrow)
>       val gradients = new ListBuffer[Array[Float]]
>       for (i <- 0 until nrow){
>         val sigmoid_pred = transPredicts(i)(0)
>         val g1 = sigmoid_pred * (1 - sigmoid_pred)
>         val g2 = label(i) + math.pow(-1, label(i)) * sigmoid_pred
>         val g3 = sigmoid_pred + label(i) - 1
>         val g4 = 1 - g2
>         val grad = (alpha*label(i) + (1 - alpha)*(1 - label(i)))*(gamma_indct * g3 * math.pow(g2, gamma_indct) * math.log(g4 + 1e-9) + math.pow(-1, label(i)) * math.pow(g2, gamma_indct + 1))
>         val hess_1 = math.pow(g2, gamma_indct) + gamma_indct * Math.pow(-1, label(i)) * g3 * math.pow(g2, gamma_indct - 1)
>         val hess_2 = math.pow(-1, label(i)) * g3 * math.pow(g2, gamma_indct) / g4
>         val hess = (alpha*label(i) + (1 - alpha)*(1 - label(i)))*(((hess_1 * math.log(g4 + 1e-9) - hess_2) * gamma_indct + (gamma_indct + 1) * math.pow(g2, gamma_indct)) * g1)
>         gradArr(i) = grad.toFloat
>         hessArr(i) = hess.toFloat
>       }
>       gradients += gradArr
>       gradients += hessArr
>       gradients.toList
>     }
>   }
> ```

Have you solved the problem? I have met the same problem.
